libraries/AP_Param/AP_Param.cpp

```diff
@@ -2662,11 +2676,258 @@ void AP_Param::show_all(AP_HAL::BetterStream *port, bool showKeyValues)
+            info.name = _empty_string;
+            return false;
+        }
+        // fill in footer for all entries
+        for (uint8_t gi=1; gi<num_params+2; gi++) {
+            auto &ginfo = const_cast<GroupInfo*>(info.group_info)[gi];
+            ginfo.name = _empty_string;
+            ginfo.idx = 0xff;
+        }
+        // hidden first parameter containing AP_Int32 crc
+        auto &hinfo = const_cast<GroupInfo*>(info.group_info)[0];
+        hinfo.flags = AP_PARAM_FLAG_HIDDEN;
+        hinfo.name = _empty_string;
+        hinfo.idx = 0;
+        hinfo.offset = 0;
+        hinfo.type = AP_PARAM_INT32;
+        // fill in default value with the CRC. Relies on sizeof crc == sizeof float
+        memcpy((uint8_t *)&hinfo.def_value, (const uint8_t *)&crc, sizeof(crc));
+    }
+
+    // remember the table size
+    if (_dynamic_table_sizes[i] == 0) {
+        _dynamic_table_sizes[i] = num_params;
+    }
+    
+    // make the group active
+    info.key = key;
+    info.type = AP_PARAM_GROUP;
+
+    invalidate_count();
+
+    // save the CRC
+    AP_Int32 *crc_param = const_cast<AP_Int32 *>((AP_Int32 *)info.ptr);
+    crc_param->set(crc);
+    crc_param->save(true);
+
+    return true;
+}
+
+/*
+ Load an AP_Int32 variable from EEPROM using top level key and group element. Used to confirm
+ a key in add_table()
+*/
+bool AP_Param::load_int32(uint16_t key, uint32_t group_element, int32_t &value)
+{
+    struct Param_header phdr;
+
+    phdr.type = AP_PARAM_INT32;
+    set_key(phdr, key);
+    phdr.group_element = group_element;
+
+    // scan EEPROM to find the right location
+    uint16_t ofs;
+    if (!scan(&phdr, &ofs)) {
+        return false;
+    }
+
+    // found it
+    _storage.read_block(&value, ofs+sizeof(phdr), type_size(AP_PARAM_INT32));
+    return true;
+}
+
+/*
+  add a parameter to a dynamic table
+ */
+bool AP_Param::add_param(uint8_t _key, uint8_t param_num, const char *pname, float default_value)
+{
+    // check for valid values
+    if (param_num == 0 || param_num > 63 || strlen(pname) > 16) {
+        return false;
+    }
+
+    uint16_t key = uint16_t(_key) + AP_PARAM_DYNAMIC_KEY_BASE;
+    // find the info
+    uint8_t i;
+    for (i=0; i<AP_PARAM_MAX_DYNAMIC; i++) {
+        auto &info = _var_info_dynamic[i];
+        if (info.key == key) {
+            break;
+        }
+    }
+    if (i == AP_PARAM_MAX_DYNAMIC) {
+        // not found
+        return false;
+    }
+
+    if (param_num > _dynamic_table_sizes[i]) {
+        return false;
+    }
+
+    auto &info = _var_info_dynamic[i];
+    if (info.ptr == nullptr) {
+        return false;
+    }
+
+    // check CRC
+    auto &hinfo = const_cast<GroupInfo*>(info.group_info)[0];
+    const int32_t crc = *(const int32_t *)(&hinfo.def_value);
+
+    int32_t current_crc;
+    if (load_int32(key, 0, current_crc) && current_crc != crc) {
+        // crc mismatch, we have a conflict with an existing use of this key
+        return false;
+    }
```

Comment:
What case does this protects against? The table should not be loaded in the first place with a invalid crc?

Reply:
needed for two scripts with same key